src/site/content/en/blog/asyncify/index.md

```diff
@@ -0,0 +1,516 @@
+if (mode == REWINDING) {
+    restoreLocals();
+    mode = NORMAL_EXECUTION;
+}
+puts("B");
+```
+
+Initially `mode` is set to `NORMAL_EXECUTION`. Correspondingly, the first time such transformed code
+is executed, only the part leading up to "async_sleep" will get evaluated. As soon as the
+asynchronous operation is scheduled, Asyncify saves all the locals, and unwinds the stack by
+returning from each function all the way to the top, this way giving control back to the browser
+event loop.
+
+Then, once `async_sleep()` resolves, Asyncify support code will change `mode` to `REWINDING`, and
+call our function again. This time, the "normal execution" branch is skipped - since it already did
+the job last time and we want to avoid printing "A" twice - and instead it comes straight to the
+"rewinding" branch. Once it's reached, it restores all the stored locals, changes mode back to
+"normal" and continues the execution as if the code was never stopped in the first place.
+
+### Transformation costs
+
+Unfortunately, Asyncify transform isn't completely free, since it has to inject quite a bit of
+supporting code for storing and restoring all those locals, navigating the call stack under
+different modes and so on. It tries to modify only functions marked as asynchronous on the command
+line, as well as any of their potential callers, but that still might result in ~50% overhead to the
+uncompressed code size.
+
+{% Img src="image/9oK23mr86lhFOwKaoYZ4EySNFp02/Im4hmQOYRHFcsg8UTfTR.png", alt="A graph showing code size overhead for various benchmarks, from near-0% under fine-tuned conditions to over 100% in worst cases",
+width="800", height="494" %}
+
+It's not ideal, but in many cases acceptable when the alternative is not having the functionality
+altogether or having to make significant rewrites to the original code.
+
+Make sure to always enable optimizations for the final builds to avoid it going even higher. You can
+also check [Asyncify-specific optimization
+options](https://emscripten.org/docs/porting/asyncify.html#optimizing) to reduce the overhead by
+limiting transforms only to specified functions and/or only direct function calls. There is also a
+minor cost to runtime performance, but it's limited to the async calls themselves. However, compared
+to the cost of the actual work, it's usually negligible.
+
+## Real-world demos
+
+Now that we've looked at the simple examples, let's move on to more complicated scenarios.
+
+As mentioned in the beginning of the article, one of the storage options on the Web is an
+asynchronous [File System Access API](https://web.dev/file-system-access/). It provides access to a
+real host filesystem from a Web application.
+
+On the other hand, there is a de-facto standard—[WASI](https://github.com/WebAssembly/WASI) - for
+WebAssembly I/O in the console and the server-side. It was designed as a compilation target for
+system languages, and exposes all sorts of filesystem and other operations in a traditional
```

Comment:
```suggestion
system languages, and exposes all sorts of file system and other operations in a traditional
```

Reply:
Hm, isn't filesystem commonly written as a single word?

Reply:
I'm not aware of a standard, but two words is more common web.dev than one word by about 10 to 1. Please change.